### PR TITLE
support systemd cgroups

### PR DESCRIPTION
**Pull request by dustin-bingham**

### CODE REVIEW

File: util/string_view.inl [hunk 1]
```diff
@@ -141,6 +141,15 @@ inline std::string_view slice_suffix(std::string_view view, std::size_t size)
   return view;
 }
 
+inline bool starts_with(std::string_view view, std::string_view prefix)
+{
+  if (view.length() < prefix.length()) {
+    return false;
+  }
```
Comment (yonch) [inline util/string_view.inl:148]: Seems `substr` already handles this case gracefully:

```
Returns a view of the substring [pos, pos + rcount), where rcount is the smaller of count and size() - pos.
```
from https://en.cppreference.com/w/cpp/string/basic_string_view/substr

So consider replacing this check with a "// note: view.substr() already handles the case where view.length() > prefix.length()". or something similar (this should eliminate the run-time cost of the redundant check)

File: util/string_view.h [hunk 2]
```diff
@@ -175,6 +175,11 @@ std::string_view slice_suffix(std::string_view view, std::size_t size);
  */
 bool ends_with(std::string_view view, std::string_view suffix);
 
+/**
+ * Returns whether the view starts with the specified prefix.
+ */
+bool starts_with(std::string_view view, std::string_view prefix);
+
```
Comment (yonch) [inline util/string_view.h:182]: nit. C++20 string_view supports `starts_with`. At this point we could consider requiring a C++20 compiler..

File: render/npm.render [hunk 3]
```diff
@@ -415,7 +415,7 @@ app ingest {
       severity 0
       1: u64 cgroup
       2: u64 cgroup_parent
-      3: u8 name[64]
+      3: u8 name[256]
```
Comment (yonch) [inline render/npm.render:418]: this message is sent from kernel collector to reducer (in `buffered_poller.cc`).

unfortunately, the protocol currently does not support forwards and backwards compatibility (we implemented most of the mechanism but had not put everything together..).

what we have been doing so far when messages change (and we should consider doing here), is to create a new message with a new RPC ID that has a new format. Then, the old handler calls the new handler after translating the message. For example `pid_info_create_deprecated` and `pid_info_create`.

File: util/string_view.inl (hunk 1 above)
Comment (dustin-bingham) [reply util/string_view.inl]: done

File: util/string_view.h (hunk 2 above)
Comment (dustin-bingham) [reply util/string_view.h:182]: yes, i agree - would be nice to move ahead to c++ 20.

File: render/npm.render (hunk 3 above)
Comment (dustin-bingham) [reply render/npm.render]: done (I think - are the rpc ids scoped, or do I need to avoid number collisions throughout the entire file?)

**Review by yonch — APPROVED** [review]
> done (I think - are the rpc ids scoped, or do I need to avoid number collisions throughout the entire file?) 

rpc ids are scoped per app. need to verify there was a sufficient pool of IDs assigned to the app (which I believe there were; if I didn't mis-count, there should be 113 assigned IDs which would support up to ID 112)

**Comment by yonch** [comment]
> done (I think - are the rpc ids scoped, or do I need to avoid number collisions throughout the entire file?) 

rpc ids are scoped per app. need to verify there was a sufficient pool of IDs assigned to the app (which I believe there were; if I didn't mis-count, there should be 113 assigned IDs which would support up to ID 112)

**Pull request merged**